src/order/jordan_hoelder.lean

```diff
@@ -0,0 +1,803 @@
+    { intro i,
+      rw [fin.snoc_cast_succ, ← fin.cast_succ_fin_succ, fin.snoc_cast_succ],
+      exact s.step _ }
+  end }
+
+@[simp] lemma top_snoc
+  (s : composition_series X)
+  (x : X)
+  (hsat : is_maximal s.top x) :
+  (snoc s x hsat).top = x :=
+fin.snoc_last _ _
+
+@[simp] lemma snoc_last
+  (s : composition_series X)
+  (x : X)
+  (hsat : is_maximal s.top x) :
+  snoc s x hsat (fin.last (s.length + 1)) = x :=
+fin.snoc_last _ _
+
+@[simp] lemma snoc_cast_succ
+  (s : composition_series X)
+  (x : X)
+  (hsat : is_maximal s.top x) (i : fin (s.length + 1)) :
+  snoc s x hsat (i.cast_succ) = s i :=
+fin.snoc_cast_succ _ _ _
+
+@[simp] lemma bot_snoc
+  (s : composition_series X)
+  (x : X)
+  (hsat : is_maximal s.top x) :
+  (snoc s x hsat).bot = s.bot :=
+by rw [bot, bot, ← fin.cast_succ_zero, snoc_cast_succ]
+
+lemma mem_snoc
+  {s : composition_series X}
+  {x y: X}
+  {hsat : is_maximal s.top x} :
+  y ∈ snoc s x hsat ↔ y ∈ s ∨ y = x :=
+begin
+  simp only [snoc, mem_def],
+  split,
+  { rintros ⟨i, rfl⟩,
+    refine fin.last_cases _ (λ i, _) i,
+    { right, simp },
+    { left, simp } },
+  { intro h,
+    rcases h with ⟨i, rfl⟩ | rfl,
+    { use i.cast_succ, simp },
+    { use (fin.last _), simp } }
+end
+
+lemma eq_snoc_erase_top
+  {s : composition_series X}
+  (h : 0 < s.length) :
+  s = snoc (erase_top s) s.top
+    (is_maximal_erase_top_top h) :=
+begin
+  ext x,
+  simp [mem_snoc, mem_erase_top h],
+  by_cases h : x = s.top; simp [*, s.top_mem]
+end
+
+@[simp] lemma snoc_erase_top_top {s : composition_series X}
+  (h : is_maximal s.erase_top.top s.top) :
+  s.erase_top.snoc s.top h = s :=
+have h : 0 < s.length,
+  from nat.pos_of_ne_zero begin
+    assume hs,
+    refine ne_of_gt (lt_of_is_maximal h) _,
+    simp [top, fin.ext_iff, hs]
+  end,
+(eq_snoc_erase_top h).symm
+
+/-- Two `composition_series X`, `s₁` and `s₂` are equivalent if there is a bijection
+`e : fin s₁.length ≃ fin s₂.length` such that for any `i`,
+`iso (s₁ i) (s₁ i.succ) (s₂ (e i), s₂ (e i.succ))` -/
+def equivalent (s₁ s₂ : composition_series X) : Prop :=
+∃ f : fin s₁.length ≃ fin s₂.length,
+  ∀ i : fin s₁.length,
+    iso (s₁ i.cast_succ, s₁ i.succ)
+    (s₂ (f i).cast_succ, s₂ (f i).succ)
+
+namespace equivalent
+
+@[refl] lemma refl (s : composition_series X) : equivalent s s :=
+⟨equiv.refl _, λ _, iso_refl_of_is_maximal (s.step _)⟩
+
+@[symm] lemma symm {s₁ s₂ : composition_series X} (h : equivalent s₁ s₂) :
+  equivalent s₂ s₁ :=
+⟨h.some.symm, λ i, iso_symm (by simpa using h.some_spec (h.some.symm i))⟩
+
+@[trans] lemma trans {s₁ s₂ s₃ : composition_series X}
+  (h₁ : equivalent s₁ s₂)
+  (h₂ : equivalent s₂ s₃) :
+  equivalent s₁ s₃ :=
+⟨h₁.some.trans h₂.some, λ i, iso_trans (h₁.some_spec i) (h₂.some_spec (h₁.some i))⟩
+
+lemma append
+  {s₁ s₂ t₁ t₂ : composition_series X}
+  (hs : s₁.top = s₂.bot)
+  (ht : t₁.top = t₂.bot)
+  (h₁ : equivalent s₁ t₁)
+  (h₂ : equivalent s₂ t₂) :
+  equivalent (append s₁ s₂ hs) (append t₁ t₂ ht) :=
```

Comment:
It seems the PR still has a number of lemmas that aren't used in the main result. This one seems reasonable, I guess, but others I found like `lt_top_of_mem_erase_top` maybe aren't going to useful elsewhere and could be removed? Not sure.

Reply:
This is the main question about this PR. I don't have a clear idea about how Jordan-Hölder is used so I don't know how useful this stuff is and I think I'll leave that call to someone who knows more about how it's used.

Reply:
I think it's fine to keep what is there. I've read through this PR, and it all looks solid. We'll see what people need later.